### PR TITLE
[solana-program] Move bn254 logic from `solana-program` into a separate crate

### PR DESCRIPTION
**Pull request by samkim-crypto**
#### Problem
The bn128 logic in `solana-program` is quite independent of the rest of the crate. Currently, the downstream projects that depend on only `solana-program`, but not bn128, will need to build with unnecessary bn128 logic, which is quite heavy.

#### Summary of Changes
Move bn128 logic into its separate crate.

Note: I named the crate to be solana-bn254. Bn254 is a more modern and descriptive name for bn128, which does not provide 128 bits of security any more. I will limit this PR to just the restructuring of the crate; I will update the contents inside the crate to use bn254 in a follow-up PR.

This is part of https://github.com/anza-xyz/agave/issues/983.

Fixes #

### CODE REVIEW

**Comment by yihau** [comment]
hey, @dmakarov! do you think it is a good idea to bump this number again in this PR? https://github.com/anza-xyz/agave/blob/f423d5b6858736ec13813d975b558a51d64acf7c/ci/test-stable.sh#L96

we got `Regression of build redundancy` error again 😢

**Comment by dmakarov** [comment]
> hey, @dmakarov! do you think it is a good idea to bump this number again in this PR?
> 
> https://github.com/anza-xyz/agave/blob/f423d5b6858736ec13813d975b558a51d64acf7c/ci/test-stable.sh#L96
> 
> we got `Regression of build redundancy` error again 😢

I think this check should be removed. It’s not useful because we don’t have a mechanism to keep the redundancy to a minimum.

**Comment by samkim-crypto** [comment]
Hm, okay, then since PR is a couple weeks old now, let me bump the threshold this time to satisfy the CI 🙏 . Let's discuss whether we want to remove the redundancy build check on a follow-up.

File: sdk/program/src/lib.rs [hunk 1]
```diff
@@ -471,7 +471,6 @@ extern crate self as solana_program;
 
 pub mod account_info;
 pub mod address_lookup_table;
-pub mod alt_bn128;
```
Comment (Lichtso) [inline sdk/program/src/lib.rs:474]: Um, what about gentle deprecation?
Maybe an alias to the new crate?

Comment (samkim-crypto) [reply sdk/program/src/lib.rs:474]: Oh right, good idea 😅

Comment (samkim-crypto) [reply sdk/program/src/lib.rs:474]: Actually, it seems difficult to import the new crate in `solana-program` due to circular dependency. It seems best to just leave the original logic in `solana-program` and add deprecation warnings pointing to the new crate. Wdyt?

Comment (kevinheavey) [reply sdk/program/src/lib.rs:474]: When https://github.com/anza-xyz/agave/pull/1827 lands, the circular dependency can be avoided by just moving the required syscall definitions from solana-program to solana-bn254

Comment (kevinheavey) [reply sdk/program/src/lib.rs:474]: #1827 has landed so you can just move the `define_syscall!` invocation and re-export the syscall like in #1656

**Review by Lichtso — DISMISSED** [review]

**Review by Lichtso — APPROVED** [review]

**Pull request merged**